File: machine/cpp/rompack/host_system_atlas.cpp
#include "rompack/host_system_atlas.h"

#include <stdexcept>
#include <string>

namespace bmsx {

const HostSystemAtlasImage& hostSystemAtlasImage(std::string_view id) {
	size_t first = 0u;
	size_t last = HOST_SYSTEM_ATLAS.images.size();
	while (first < last) {
		const size_t middle = (first + last) >> 1u;
		const HostSystemAtlasImage& image = HOST_SYSTEM_ATLAS.images[middle];
		const int order = image.id.compare(id);
		if (order < 0) {
			first = middle + 1u;
		} else if (order > 0) {
			last = middle;
		} else {
			return image;
		}
	}
	throw std::runtime_error("Image '" + std::string(id) + "' is not in the host system atlas.");
}

} // namespace bmsx

File: machine/ts/rompack/host_system_atlas.ts
import { HOST_SYSTEM_ATLAS } from './host_system_atlas.generated';

export type HostSystemAtlasImage = Readonly<{
	id: string;
	width: number;
	height: number;
	u: number;
	v: number;
	w: number;
	h: number;
}>;

export type HostSystemAtlas = Readonly<{
	width: number;
	height: number;
	pixels: Uint8Array;
	images: readonly HostSystemAtlasImage[];
}>;

export { HOST_SYSTEM_ATLAS };

export function hostSystemAtlasImage(id: string): HostSystemAtlasImage {
	let first = 0;
	let last = HOST_SYSTEM_ATLAS.images.length;
	while (first < last) {
		const middle = (first + last) >>> 1;
		const image = HOST_SYSTEM_ATLAS.images[middle];
		if (image.id < id) {
			first = middle + 1;
		} else if (image.id > id) {
			last = middle;
		} else {
			return image;
		}
	}
	throw new Error(`Image '${id}' is not in the host system atlas.`);
}
